Automattic/newspack-plugin
# feat: add ga4 user registered handler

### All Submissions:






### Changes proposed in this Pull Request:



Closes # .

### How to test the changes in this Pull Request:

Get the GA4 credentials. You will need your measurement ID and an API secret.

* api_secret - Required. An API SECRET generated in the Google Analytics UI. To create a new secret, navigate to:
    Admin > Data Streams > choose your stream > Measurement Protocol > Create
* measurement_id - Required. The measurement ID associated with a stream. Found in the Google Analytics UI under:
    Admin > Data Streams > choose your stream > Measurement ID

Store this info in the database:
```
wp option set ga4_measurement_id "G-XXXXXXXXXX"
wp option set ga4_api_secret YYYYYYYYYYYYYYYYYYY
```
Now, let's test:

1. Make sure RAS is active and that you have a registered reader
2. Make sure to add this constant to wp-config
```
define( 'NEWSPACK_LOG_LEVEL', 2 );
define( 'NEWSPACK_EXPERIMENTAL_GA4_EVENTS', true );
```
4. Register the site (either via a registration prompt or from the my account page)
5. Watch the `reader_registered` event being fired in the logs
```
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::dispatch]: Dispatching action "reader_registered" via Action Scheduler.
```
6. Watch the handler being triggered and sending the `reader_registered` GA event:
```
[NEWSPACK-DATA-EVENTS][Newspack\Data_Events::handle]: Executing action handlers for "reader_registered".
[NEWSPACK-DATA-EVENTS-GA4][Newspack\Data_Events\Connectors\GA4::log]: Event sent - reader_registered
```
7. Visit the Google Analytics Realtime dashboard
8. See the `reader_registered` events coming in
9. Inspect them and see the `registration_method` attribute set as `registration-block` for the times you registered via the registration block.
10. See that the `referer` attribute is correct
11. Register from inside a popup and see that, in those cases, you'll also get some campaign specific data:
```
(
    [logged_in] => yes
    [ga_session_id] => 1677185108
    [is_reader] => yes
    [registration_method] => registration-block-popup
    [campaign_id] => 139
    [campaign_title] => Registration Overlay
    [campaign_frequency] => always
    [campaign_placement] => center
    [campaign_has_registration_block] => 1
    [campaign_has_donation_block] => 
    [campaign_has_newsletter_block] => 
    [referer] => /2023/02/07/accumsan-nam-mauris-erat-ultrices-nascetur-porta-ornare-est-class-cursus-massa-lectus/
)

```
(you can also add a debug to the `handle_reader_registered` method to see these params being set).

Note that this PR also fixes kind of a bug. Something that was working by accident.

In the line:
```PHP
$metadata['current_page_url']    = home_url( add_query_arg( array(), \wp_get_referer() ) );
```

`wp_get_referer()` always returns `false` because the referer is the same page as the current page. (It's a post to itself). The accident is that if the second parameter to `add_query_arg` is `false`, it will use the current URL instead. And since this is a post to itself, the result is the same...

But my code fixes it and we now actually read the value from the `_wp_http_referer` in the form.

### Other information:
